src/Processors/Transforms/buildPushingToViewsChain.cpp

```diff
@@ -274,6 +274,9 @@ Chain buildPushingToViewsChain(
             running_group = current_thread->getThreadGroup();
         else
             running_group = std::make_shared<ThreadGroupStatus>();
+        /// Always set query_context for system.query_views_log
```

Comment:
TBH, I don't remember why I never implemented `query_views_log` for background inserts. I remember testing it and we have such views in production, but I don't remember why I decided to not work on it. Thanks!

Will this play well with https://github.com/ClickHouse/ClickHouse/pull/40732/? (It was your fix too :heart:)
Or worded better: will the memory / objects attached to that context be freed at the end of the MV execution or will they live forever attached to the main context?

Reply:
>will the memory / objects attached to that context be freed at the end of the MV execution or will they live forever attached to the main context?

Yes, it should be OK.
Since there was a problem with global `ThreadGroupStatus` (`MainThreadStatus::getInstance().getThreadGroup()`)

Reply:
Ok, now I remember one of the potential problems for this.

If you create a MV on top of system.query_views_log (for aggregation/analysis later) then you will have new logs constantly:

```
CREATE TABLE a
(
    `a` Int64
)
ENGINE = MergeTree
ORDER BY a;

CREATE TABLE b AS a;

CREATE MATERIALIZED VIEW a_to_b TO b AS
SELECT *
FROM a;

-- Very stupid MV to showcase the example
CREATE MATERIALIZED VIEW query_views_log_to_b TO b AS
SELECT count()
FROM system.query_views_log;

INSERT INTO a SELECT *
FROM numbers(100);
```


This will create an infinite number of rows (it will keep creating new rows related to the MV from query_views_log which would trigger a new row):

```
Row 1:
──────
event_date:              2023-02-22
event_time:              2023-02-22 10:20:13
event_time_microseconds: 2023-02-22 10:20:13.587659
view_duration_ms:        0
initial_query_id:        77a42294-bf5e-46bd-bc54-502017df4074
view_name:               default.a_to_b
view_uuid:               0a88273e-fe84-4796-81c6-c19e87d2756f
view_type:               Materialized
view_query:              SELECT * FROM default.a
view_target:             default.b
read_rows:               100
read_bytes:              800
written_rows:            100
written_bytes:           800
peak_memory_usage:       0
ProfileEvents:           {'FileOpen':7,'WriteBufferFromFileDescriptorWrite':7,'WriteBufferFromFileDescriptorWriteBytes':741,'IOBufferAllocs':14,'IOBufferAllocBytes':3242866,'DiskWriteElapsedMicroseconds':24,'InsertedRows':100,'InsertedBytes':800,'SelectedRows':100,'SelectedBytes':800,'MergeTreeDataWriterRows':100,'MergeTreeDataWriterUncompressedBytes':800,'MergeTreeDataWriterCompressedBytes':497,'MergeTreeDataWriterBlocks':1,'MergeTreeDataWriterBlocksAlreadySorted':1,'InsertedCompactParts':1,'ContextLock':29,'RWLockAcquiredReadLocks':4,'RealTimeMicroseconds':610,'SoftPageFaults':8}
status:                  QueryFinish
exception_code:          0
exception:               
stack_trace:             

Row 2:
──────
event_date:              2023-02-22
event_time:              2023-02-22 10:20:18
event_time_microseconds: 2023-02-22 10:20:18.768816
view_duration_ms:        0
initial_query_id:        
view_name:               default.query_views_log_to_b
view_uuid:               6d5ecb51-9d58-4c3c-a0af-1d122682d41b
view_type:               Materialized
view_query:              SELECT count() FROM system.query_views_log
view_target:             default.b
read_rows:               0
read_bytes:              0
written_rows:            1
written_bytes:           0
peak_memory_usage:       0
ProfileEvents:           {'ArenaAllocChunks':1,'ArenaAllocBytes':4096,'InsertedRows':1,'SelectedRows':1,'SelectedBytes':1,'ContextLock':23,'RWLockAcquiredReadLocks':4,'RealTimeMicroseconds':391,'SoftPageFaults':17}
status:                  QueryFinish
exception_code:          0
exception:               
stack_trace:             

Row 3:
──────
event_date:              2023-02-22
event_time:              2023-02-22 10:20:26
event_time_microseconds: 2023-02-22 10:20:26.270970
view_duration_ms:        0
initial_query_id:        
view_name:               default.query_views_log_to_b
view_uuid:               6d5ecb51-9d58-4c3c-a0af-1d122682d41b
view_type:               Materialized
view_query:              SELECT count() FROM system.query_views_log
view_target:             default.b
read_rows:               0
read_bytes:              0
written_rows:            1
written_bytes:           0
peak_memory_usage:       0
ProfileEvents:           {'ArenaAllocChunks':1,'ArenaAllocBytes':4096,'InsertedRows':1,'SelectedRows':1,'SelectedBytes':1,'ContextLock':23,'RWLockAcquiredReadLocks':4,'RealTimeMicroseconds':404,'SoftPageFaults':7}
status:                  QueryFinish
exception_code:          0
exception:               
stack_trace:             

Row 4:
──────
event_date:              2023-02-22
event_time:              2023-02-22 10:20:33
event_time_microseconds: 2023-02-22 10:20:33.773233
view_duration_ms:        0
initial_query_id:        
view_name:               default.query_views_log_to_b
```

I'm not sure how how to best address this to avoid recursive inserts, or maybe it's not necessary to prevent them.

Reply:
There is the same issue for `system.text_log`, each flushing to `text_log` creates few messages that will be flushed again after, so I don't think it is a problem (anyway I guess there will be almost 0 users for MV over `system.query_views_log` and if so, this should be understandable)

Plus this is not only about `system.*_log` tables but also about Buffer flushes and so on.